packages/graphback-codegen-offix/src/OffixPlugin.ts

```diff
@@ -0,0 +1,99 @@
+import { SchemaComposer } from 'graphql-compose';
+
+/**
+ * Configuration for Schema generator CRUD plugin
+ */
+export interface OffixPluginConfig {
+
+    /**
+     * RelativePath for the output files created by generator
+     */
+    outputPath: string
+
+
+    /*
+     * RelativePath for the output files created by generator
+     */
+    // TODO separate annotation for delta queries should be used for more granular support
+    generateDeltaQueries: boolean
+
+    /**
+     * Delta resolvers format
+     */
+    deltaResolverFormat: 'ts' | 'js' | 'graphql',
+
+}
+
+export const SCHEMA_CRUD_PLUGIN_NAME = "SchemaCRUD";
+
+/**
+ * Ofix plugin
+ * 
+ * Plugin is enabled by """ @datasync """ annotation
+ * It will add version field to the schema and also generate diffQueries
+ */
+export class OffixPlugin extends GraphbackPlugin {
+
+    private pluginConfig: OffixPluginConfig;
+
+    public constructor(pluginConfig?: OffixPluginConfig) {
+        super()
+        this.pluginConfig = Object.assign({ format: 'graphql', outputFileName: 'schema' }, pluginConfig);
+        if (!pluginConfig.outputPath) {
+            throw new Error("schema plugin requires outputPath parameter")
+        }
+    }
+
+    public transformSchema(metadata: GraphbackCoreMetadata): GraphQLSchema {
+        const schema = metadata.getSchema()
+        const schemaComposer = new SchemaComposer(schema);
+        const models = metadata.getModelDefinitions();
+        if (models.length === 0) {
+            this.logWarning("Provided schema has no models. Returning original schema without any changes.")
+
+            return schema;
+        }
+
+        models.forEach((model: ModelDefinition) => {
+            // TODO use `versioned` marker to check if we should add version
+            const modifiedType = schemaComposer.getOTC(model.graphqlType.name);
+            modifiedType.addFields({
+                version: 'Int'
+            });
+
+            const inputType = schemaComposer.getITC(getInputTypeName(model.graphqlType.name))
+            if (inputType) {
+                inputType.addFields({
+                    version: 'Int'
+                });
+            }
+
+            // Diff queries
+            if (this.pluginConfig.generateDeltaQueries) {
+                const diffQuery = `${model.graphqlType.name  }Delta`
+                schemaComposer.Query.addFields({
+                    [diffQuery]: `[${model.graphqlType.name}]!`
+                })
+                schemaComposer.Query.addFieldArgs(diffQuery, {
+                    lastSync: 'String!'
+                })
+                // TODO generate delta resolvers
+            }
+        })
+
+        return schemaComposer.buildSchema()
+    }
+
+    public createResources(metadata: GraphbackCoreMetadata): void {
```

Comment:
Can this be removed completely if it is not used?

Reply:
Sadly not. It is defined as abstract in the top class additionally it will be used when I add resolver generation for diff queries. This plugin is just to get us sense how integration will work. We will have it refined later on.

Reply:
I could get that as non abstract but then when implementing the class there will be no methods created - there needs to be some abstract method.